references/debugging_guide.md

```diff
@@ -0,0 +1,95 @@
+   5. Add a test to the suite of regression tests to reproduce this specific software failure instance. Check the test in with git.
+      - See chapter 4, "Reproducing Problems" of "Why Programs Fail: A Guide to Systematic Debugging" by Morgan Kaufmann.
+   6. Create an automated and simplified test case for this specific failure instance.
+      - See chapter 5, "Simplifying Problems" of "Why Programs Fail: A Guide to Systematic Debugging" by Morgan Kaufmann.
+
+2. Understand how the system that is producing the failure is supposed to work.
+   1. Identify and read the basic documentation related to the (relevant?) external software tools being used at or around the point of failure.
+   2. Take basic trainings (e.g. online tutorials, online courses, etc.) in the external tools being used in your project.
+   3. Identify, store, and read the details about major / complex algorithms that one's software is implementing.
+
+3. Understand how the system that is producing the failure is working.
+   1. What is being produced during failure?
+
+      Ensure that one has a debugging system in-place to identify and record the intermediate values being produced by one's software as well as which of those values are incorrect in the simplest created test case.
+
+   2. When does failure occur?
+
+      If failure is intermittent or conditional, document what inputs, if any, cause the failure to occur rather than not.
+
+   3. Where is the defective code?
+      1. Hypothesize and record possible points where the bug could have been introduced.
+         1. For each hypothesis, note its predicted observations. These observations will be tested for when attempting to falsify the hypothesis.
+         2. For each hypothesis, note the empirical facts that suggest it.
+         3. Set a time limit for this data and test-less work.
+         4. Be sure to include logically likely contributors to the failure's existence (e.g. other known bugs, causes in program state / code / input, anomolies in the output or intermediate variables, poorly written code, etc.).
+      2. For each of those hypothesized points, determine whether the point is defective.
+      3. For each defective point, determine whether that point actually changes the failure behavior.
+      4. Employ various brute-force / common search techniques to discover remaining causes of the failure. E.g.: [WIP--to be detailed.]
+         1. Bisection Search
+            1. Start at the beginning of a chunk of code where there is no code failure.
+            2. Mark the earliest point in the code chunk where the failure has occurred.
+            3. Go halfway between the two points and identify whether there are any infections (incorrect states) in the first half of the code chunk.
+            4. If infections are found, repeat steps 1 to 3 on the code chunk between the original starting point and the earliest new infection that was found.
```

Comment:
```suggestion
            4. If infections are found, repeat steps a to c on the code chunk between the original starting point and the earliest new infection that was found.
```
In the print out of the markdown, these sub-bullets are lettered. Just a formatting issue in the Markdown file.

Reply:
Changed!